src/openrct2/management/Research.cpp

```diff
@@ -910,3 +890,107 @@ bool ResearchItem::Exists() const
+        return;
+
+    if (researchItem->type != RESEARCH_ENTRY_TYPE_RIDE)
+        return;
+
+    auto rideType = researchItem->baseRideType;
+    if (rideType >= RIDE_TYPE_COUNT)
+        return;
+
+    const auto& rtd = RideTypeDescriptors[rideType];
+    if (rtd.HasFlag(RIDE_TYPE_FLAG_LIST_VEHICLES_SEPARATELY))
+    {
+        researchItem->flags |= RESEARCH_ENTRY_FLAG_FIRST_OF_TYPE;
+        return;
+    }
+
+    if (!rtd.HasFlag(RIDE_TYPE_FLAG_HAS_RIDE_GROUPS))
+    {
+        if (!_seenRideType[rideType])
+            researchItem->flags |= RESEARCH_ENTRY_FLAG_FIRST_OF_TYPE;
+
+        _seenRideType[rideType] = true;
+    }
+    else
+    {
+        const auto& entry = get_ride_entry(researchItem->entryIndex);
+        if (entry != nullptr)
+        {
+            auto rideGroupIndex = RideGroupManager::GetRideGroupIndex(rideType, entry);
+            assert(rideGroupIndex < MAX_RIDE_GROUPS_PER_RIDE_TYPE);
+
+            if (!_seenRideGroup[rideType * rideGroupIndex])
+                researchItem->flags |= RESEARCH_ENTRY_FLAG_FIRST_OF_TYPE;
+
+            _seenRideGroup[rideType * rideGroupIndex] = true;
+        }
+    }
+}
+
+void research_determine_first_of_type()
+{
+    _seenRideType.reset();
+    _seenRideGroup.reset();
+
+    for (auto& researchItem : gResearchItemsInvented)
+    {
+        if (researchItem.type != RESEARCH_ENTRY_TYPE_RIDE)
+            continue;
+
+        auto rideType = researchItem.baseRideType;
+        if (rideType >= RIDE_TYPE_COUNT)
+            continue;
+
+        const auto& rtd = RideTypeDescriptors[rideType];
+        if (rtd.HasFlag(RIDE_TYPE_FLAG_LIST_VEHICLES_SEPARATELY))
+            continue;
+
+        // The next research item is also present in gResearchItemsInvented, even though it isn't invented yet(!)
+        if (gResearchNextItem.has_value() && !gResearchNextItem->IsNull() && researchItem.Equals(&gResearchNextItem.value()))
+            continue;
+
+        if (!rtd.HasFlag(RIDE_TYPE_FLAG_HAS_RIDE_GROUPS))
+        {
+            _seenRideType[rideType] = true;
+            log_error("Seen %d", rideType);
+        }
+        else
+        {
+            const auto& entry = get_ride_entry(researchItem.entryIndex);
+            if (entry != nullptr)
+            {
+                auto rideGroupIndex = RideGroupManager::GetRideGroupIndex(rideType, entry);
+                assert(rideGroupIndex < MAX_RIDE_GROUPS_PER_RIDE_TYPE);
+                _seenRideGroup[rideType * rideGroupIndex] = true;
+                log_error("Seen rgi %d ri %d", rideGroupIndex, rideType);
+            }
+        }
+    }
+
+    if (gResearchLastItem.has_value())
+    {
+        auto tmpVal = gResearchLastItem.value();
+        research_update_first_of_type(&tmpVal);
+        gResearchLastItem = tmpVal;
+    }
+    if (gResearchNextItem.has_value())
+    {
+        auto tmpVal = gResearchNextItem.value();
+        research_update_first_of_type(&tmpVal);
+        gResearchNextItem = tmpVal;
+    }
```

Comment:
Seems like you should extract a function and call `foo(gResearchLastItem)` and `foo(gResearchNextItem)`